image/save: Derive Descriptor from actual tar archive

vvoland opened the pull request:
- fixes: https://github.com/moby/moby/issues/47072
- follow up to: https://github.com/moby/moby/pull/47068

Graphdriver layer size is the sum of the individual files size (like `du -s`), not the tar archive (`tar | wc -c`). The layer descriptor in OCI manifest specifies the byte size of the tar archive. Use the total bytes read returned by `io.Copy` to populate the `Size` field.

Also set the digest to the actual digest of the tar archive.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

vvoland committed:
image/save: Derive Descriptor from actual tar archive

Layer size is the sum of the individual files count, not the tar
archive. Use the total bytes read returned by `io.Copy` to populate the
`Size` field.

Also set the digest to the actual digest of the tar archive.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

rumpl commented on image/tarexport/save.go:
```diff
@@ -38,6 +38,7 @@ type saveSession struct {
 	images      map[image.ID]*imageDescriptor
 	savedLayers map[string]struct{}
 	diffIDPaths map[layer.DiffID]string // cache every diffID blob to avoid duplicates
+	diffIDDescs map[layer.DiffID]distribution.Descriptor
```
I'm kinda failing to understand why we need this map

vvoland replied:
We need it because `saveLayer` [doesn't read and save the tar archive if it was alread saved during this session](https://github.com/moby/moby/blob/33b5a5813b805aaba28c06ee1a8aa5b1f2435a1d/image/tarexport/save.go#L485).

So in this case we need to remember the generated descriptor and return it for any subsequent calls.

rumpl replied:
Gotcha!

rumpl commented on image/tarexport/save.go:
```diff
@@ -516,20 +522,21 @@ func (s *saveSession) saveLayer(id layer.ChainID, legacyImg image.V1Image, creat
 		}
 
 		s.diffIDPaths[l.DiffID()] = layerPath
-		s.savedLayers[legacyImg.ID] = struct{}{}
```
We don't need `savedLayers` any more?

vvoland replied:
Whoops, restored. I didn't intend to remove it in this PR (it _could_ be deleted though, but that's out of scope here).

rumpl replied:
Hm, maybe we _do_ want to remove the saved layers, there is [this test](https://github.com/moby/moby/blob/33b5a5813b805aaba28c06ee1a8aa5b1f2435a1d/image/tarexport/save.go#L454) on top, don't we want to test `diffIDDescs` at the beginning and return that desc if it exists? Right now it will return an empty desc for already saved layers

vvoland replied:
Removed unused maps and refactored the function to better distinguish between saved image (Config) and saved layer (diff tar archive).

vvoland committed:
image/save: Refactor saveLayer

`diffIDPaths` is not used and can be removed.

`savedConfig` stores if the config was already saved (ID of the image is
the ID of the config).
`savedLayers` stores if the layer (diff ID) was already saved.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

rumpl commented on image/tarexport/save.go:
```diff
@@ -36,8 +38,8 @@ type saveSession struct {
-	savedLayers map[string]struct{}
-	diffIDPaths map[layer.DiffID]string // cache every diffID blob to avoid duplicates
+	savedLayers map[layer.DiffID]distribution.Descriptor
+	savedImages map[string]struct{}
```
nit: `savedImages` is used in `saveConfig`, you might want to give these the same name

vvoland replied:
Done

thaJeztah commented:
Discussed this one on slack; this PR fixes the issue that was reported; most of the code changes are "moving" code, and this should have no direct impact on consumers of the "non-OCI" (legacy layout).


Let me bring this one in in to fix the regression 👍
